polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp

```diff
@@ -114,6 +102,97 @@ bool isCandidateLoopNest(LoopLikeOpInterface loop) {
   return true;
 }
 
+/// Get the size of unused shared local memory arena in bytes.
+unsigned getLocalMemoryRemain(gpu::GPUModuleOp &module,
+                              unsigned localMemorySize) {
+  assert(module.hasTrait<OpTrait::SymbolTable>() &&
+         "Expecting module with SymbolTable trait");
+  unsigned localMemoryRemain = localMemorySize;
+  module.walk([&localMemoryRemain](memref::GlobalOp global) {
+    MemRefType memRefTy = global.getType();
+    if (!isLocalAccessAddrSpace(memRefTy))
+      return WalkResult::advance();
+    unsigned globalSize =
+        memRefTy.getElementTypeBitWidth() * memRefTy.getNumElements() / 8;
+    if (globalSize >= localMemoryRemain) {
+      localMemoryRemain = 0;
+      return WalkResult::interrupt();
+    }
+    localMemoryRemain -= globalSize;
+    return WalkResult::advance();
+  });
+  return localMemoryRemain;
+}
+
+/// Get the require local memory for \p accTy.
+unsigned getReqdLocalMemory(sycl::AccessorType accTy,
+                            sycl::ReqdWorkGroupSize reqdWGSize) {
+  assert(!reqdWGSize.empty() && "Expecting non-empty reqdWGSize");
+  unsigned elemSize = accTy.getType().getIntOrFloatBitWidth();
+  unsigned memrefReqdLocalMemory = elemSize;
+  for (unsigned dim = 0; dim < accTy.getDimension(); ++dim) {
+    memrefReqdLocalMemory *= reqdWGSize[dim];
+  }
+  return memrefReqdLocalMemory;
+}
+
+/// Get the require local memory for memrefs in \p loopToSharedMemref.
```

Comment:
Add (i.e. for each kernel)

Reply:
Add "if there are multiple loops in the kernel that use shared local memory we return the max amount reuired by any of them."

Reply:
done.